### DISCUSSION TITLE
Vector variables don't work with distributed mesh?

### DISCUSSION THREAD

**Discussion by Eilloo** [idaholab/moose#24135] [Q&A General] [answered]
Hi all,

Are vector variables compatible with distributed meshes/Nemesis format?

The behaviour I'm seeing is the simulation solves fine, and I can open the Nemesis files in Paraview. 
However, the values of my vector variables contain garbage (e+37), and Paraview crashes if I try and go to any non-zero timestep.
The other variables look fine, and if I solve using a replicated mesh and exodus output, everything looks normal.
To be clear, I have been able to inspect my variables using another tool to see that the e+37 values persist at non-zero timesteps.

It seems like there are 3 likely possibilities: Either vector variables are not compatible, or I am consistently setting up something wrong with the input files, or there is a bug here.

Below is a MWE to reproduce this:

```
filename = NemesisMWEOut
inlet_velocity = 1e-4   #Re = 100

[Mesh]
    [base]
        type = GeneratedMeshGenerator
        dim = 2
        xmin = 0
        ymin = 0
        xmax = 10
        ymax = 2
        nx = 50
        ny = 10
    []
    allow_renumbering = false
    parallel_type = DISTRIBUTED
[]

[Variables]
    [pressure]
        family = LAGRANGE
    []
    [velocity]
        family = LAGRANGE_VEC
    []
[]

[Materials]
    [fluid]
        type = ADGenericConstantMaterial
        prop_names = 'rho mu cp k'
        prop_values = '1000 1e-3 1 1'
    []
    [ins_mat]
        type = INSADStabilized3Eqn
        velocity = velocity
        pressure = pressure
        temperature = '288'
        alpha = 1.0
    []
[]

[Kernels]
    [mass]
        type = INSADMass
        variable = pressure
    []
    [momentum_advection]
        type = INSADMomentumAdvection
        variable = velocity
    []
    [momentum_pressure]
        type = INSADMomentumPressure
        variable = velocity
        pressure = pressure
    []
    [momentum_viscous]
        type = INSADMomentumViscous
        variable = velocity
    []
    [mass_pspg]
        type = INSADMassPSPG
        variable = pressure
    []
[]

[BCs]
    [inlet]
        type = VectorDirichletBC
        variable = velocity
        values = '${inlet_velocity} 0 0'
        boundary = left
    []
    [outlet]
        type = ADDirichletBC
        variable = pressure
        value = 0
        boundary = right
    []
    [no_slip]
        type = VectorDirichletBC
        variable = velocity
        values = '0 0 0'
        boundary = 'top bottom'
    []
[]

[Preconditioning]
    [precond]
        type = SMP
        full = true
    []
[]

[Executioner]
    type = Steady
    solve_type = NEWTON
    petsc_options_iname = '-pc_type -pc_factor_shift_type'
    petsc_options_value = 'lu       NONZERO'
[]

[Outputs]
    [nemesis]
        type = Nemesis
        file_base = ${filename}
        enable = true
    []
[]
```

**Comment by lindsayad** [comment]
Distributed mesh and nemesis output are two distinct features. We test the INSAD implementation with distributed mesh every day on CIVET so I'm very confident that vector variables work with distributed meshes. However, I wouldn't be all that surprised if there are some issues with vector variables and nemesis output. @roystgnr have you ever looked at vector FE types in Nemesis?

**Reply by roystgnr** [reply]
I see a unit test for Nemesis with MONOMIAL_VEC (both Replicated and Distributed) ... and that's it.  And that goes down a slightly different code path than LAGRANGE_VEC would, so it's quite possible we have some bug in the latter that isn't tripping a unit test for the former.

**Reply by lindsayad** [reply]
@Eilloo have you run in `devel` or `dbg` mode? I am curious whether you trigger any assertions

**Reply by Eilloo** [reply]
That's encouraging to hear, and sounds like there's a chance the problem is at our end. I haven't run in devel or dbg, no - I'll do some digging there and report back

**Reply by Eilloo** [reply]
Out of interest, are you using Nemesis output file type, or something else for your own cases?

**Reply by GiudGiud** [reply]
Most of the time exodus, but for large cases on clusters we use nemesis

**Reply by Eilloo** [reply]
Hmm, the same as us... I have checked that similar behaviour occurs both locally (still using Nemesis output and distributed mesh) and on a cluster. In both instances, exodus output type without distributed mesh is fine.

I've been trying to run in dbg, but can't seem to get this working both in parallel and on a cluster, since the xterm windows can't be displayed. Right now I don't have the option to do this locally, but I'll report back if I can get it working in one way or another.

**Reply by GiudGiud** [reply]
You can run in dbg mode without drumming up xterm btw.
It s just a matter of using the dbg executable. This will let us know if the dbg version, which has plenty of additional checks built-in, errors out on vector variable outputs

**Reply by Eilloo** [reply]
Ah okay, so I basically replace the -opt file with the -dbg and see what appears in the .out? So no need to set any breakpoint flags and so forth...

**Reply by Eilloo** [reply]
The output when running as described above is the same as running the -opt (convergence info etc), with the additional information below appended:

(edit: This is using 4 processors)

```
| Reference count information                                                |
 ---------------------------------------------------------------------------- 
| N7libMesh10FEAbstractE reference count information:
|  Creations:    279
|  Destructions: 279
| N7libMesh10Parameters5ValueE reference count information:
|  Creations:    7968
|  Destructions: 7968
| N7libMesh12SparseMatrixIdEE reference count information:
|  Creations:    11
|  Destructions: 11
| N7libMesh13NumericVectorIdEE reference count information:
|  Creations:    30
|  Destructions: 30
| N7libMesh15EquationSystemsE reference count information:
|  Creations:    1
|  Destructions: 1
| N7libMesh15GhostingFunctorE reference count information:
|  Creations:    22
|  Destructions: 22
| N7libMesh15NonlinearSolverIdEE reference count information:
|  Creations:    1
|  Destructions: 1
| N7libMesh4ElemE reference count information:
|  Creations:    632
|  Destructions: 632
| N7libMesh4NodeE reference count information:
|  Creations:    561
|  Destructions: 561
| N7libMesh5QBaseE reference count information:
|  Creations:    145
|  Destructions: 145
| N7libMesh6DofMapE reference count information:
|  Creations:    2
|  Destructions: 2
| N7libMesh6SystemE reference count information:
|  Creations:    2
|  Destructions: 2
| N7libMesh9DofObjectE reference count information:
|  Creations:    1193
|  Destructions: 1193
 ---------------------------------------------------------------------------- 
```

**Reply by GiudGiud** [reply]
Thanks.
This means we are not breaking additional checks in debug mode.

**Reply by Eilloo** [reply]
A further question: What happens when I set `parallel_type = DISTRIBUTED` in the mesh block, but also set the output file type to use exodus?
I get warnings that exodus only supports serial meshes, but it goes ahead and solves, giving me an exodus file where the solution looks fine.
Does specifying `exodus` in the output block overwrite the use of a distributed mesh? Or does it somehow stitch the separate nemesis files together?

The reason is to test whether the problem is at the 'distributed mesh' stage, or if it is indeed specific to Nemesis.

**Reply by GiudGiud** [reply]
If you do that, we force a serialization then output to exodus.
it does not stitch nemesis together

**Reply by lindsayad** [reply]
To be clear we only serialize for Exodus output on processor 0. So on processor 0 you will have a copy of the entire mesh but all other processes will still only have parts of the mesh

**Reply by GiudGiud** [reply, marked as answer]
I confirmed the problem and raised this issue
https://github.com/idaholab/moose/issues/24193

We will address it when we have time. In the mean time, please just use this auxkernel to output each component separately. 
If you still name each non-vector component variable like v_x v_y v_z, Paraview will still recognize it s vector

https://mooseframework.inl.gov/source/auxkernels/VectorVariableComponentAux.html
